### PR TITLE
GH issue 28044 - Update metering troubleshooting section for not enough compute resources

### PR DESCRIPTION
**Pull request by lbarbeevargas**
GH Issue #28044 
Updates based on KCS article - https://access.redhat.com/solutions/5640401

Updates to [Not enough compute resources](https://docs.openshift.com/container-platform/4.7/metering/metering-troubleshooting-debugging.html#metering-not-enough-compute-resources_metering-troubleshooting-debugging) section of the metering Troubleshooting guide.

Preview: https://deploy-preview-30798--osdocs.netlify.app/openshift-enterprise/latest/metering/metering-troubleshooting-debugging.html#metering-not-enough-compute-resources_metering-troubleshooting-debugging

### CODE REVIEW

**Comment by lbarbeevargas** [comment]
@openshift/team-documentation @timflannagan @pruan-rht PTAL

File: modules/metering-troubleshooting.adoc [hunk 1]
```diff
@@ -10,10 +10,173 @@ A common issue with metering is pods failing to start. Pods might fail to start
...
 == Not enough compute resources
 
-A common issue when installing or running metering is lack of compute resources. Ensure that metering is allocated the minimum resource requirements described in the installation prerequisites.
+A common issue when installing or running metering is a lack of compute resources. As the cluster grows and more reports are created, the `reporting-operator` pod requires more memory. If memory usage reaches the pod limit, the cluster terminates the pod with an `OOMKilled` status. Ensure that metering is allocated the minimum resource requirements described in the installation prerequisites.
+
+NOTE: CPU usage for the `reporting-operator` pod does not increase as the cluster grows.
 
 To determine if the issue is with resources or scheduling, follow the troubleshooting instructions included in the Kubernetes document https://kubernetes.io/docs/concepts/configuration/manage-compute-resources-container/#troubleshooting[Managing Compute Resources for Containers].
 
+To troubleshoot issues due to a lack of compute resources, check the following within the `operator-metering` namespace.
```
Comment (timflannagan) [inline modules/metering-troubleshooting.adoc:19]:
```suggestion
To troubleshoot issues due to a lack of compute resources, check the following within the `openshift-metering` namespace.
```

Comment (lbarbeevargas) [reply modules/metering-troubleshooting.adoc:19]: Good catch! 👀

File: modules/metering-troubleshooting.adoc [hunk 2]
```diff
@@ -10,10 +10,173 @@ A common issue with metering is pods failing to start. Pods might fail to start
 [id="metering-not-enough-compute-resources_{context}"]
 == Not enough compute resources
 
-A common issue when installing or running metering is lack of compute resources. Ensure that metering is allocated the minimum resource requirements described in the installation prerequisites.
+A common issue when installing or running metering is a lack of compute resources. As the cluster grows and more reports are created, the `reporting-operator` pod requires more memory. If memory usage reaches the pod limit, the cluster terminates the pod with an `OOMKilled` status. Ensure that metering is allocated the minimum resource requirements described in the installation prerequisites.
+
+NOTE: CPU usage for the `reporting-operator` pod does not increase as the cluster grows.
```
Comment (timflannagan) [inline modules/metering-troubleshooting.adoc:15]: This is probably sufficient, but I think the general detail here is the metering operator does not autoscale the reporting operator based on the load in the cluster.

File: modules/metering-troubleshooting.adoc [hunk 3]
```diff
@@ -10,10 +10,173 @@ A common issue with metering is pods failing to start. Pods might fail to start
...
+node-allocatable-memory-bytes                2020-04-23T09:14:00Z   2020-08-30T05:19:00Z   2020-04-23T09:14:00Z   2020-10-14T08:01:00Z   2020-12-09T12:45:12Z   230d
+...
+pod-usage-memory-bytes                       2020-04-23T09:14:00Z   2020-08-24T20:25:00Z   2020-04-23T09:14:00Z   2020-10-09T23:31:00Z   2020-12-09T12:45:12Z   230d
+----
+
+. Check the health of the `reporting-operator` pod for a high number of pod restarts:
++
+[source,terminal]
+----
+$ oc get pods
```
Comment (timflannagan) [inline modules/metering-troubleshooting.adoc:62]: This is also sufficient if it's just a mirror of that KCS article, but we can also just filter out that reporting-operator controller Pod, by running `oc get pods -l app=reporting-operator` and the output would be a single Pod that matches that label.

File: modules/metering-troubleshooting.adoc (hunk 2 above)
Comment (lbarbeevargas) [reply modules/metering-troubleshooting.adoc:15]: 👍 I can add that clarification to the note.

File: modules/metering-troubleshooting.adoc [hunk 4]
```diff
@@ -10,10 +10,173 @@ A common issue with metering is pods failing to start. Pods might fail to start
...
+  metadata:
+  name: operator-metering
+   namespace: operator-metering
+  spec:
+   openshift-reporting:
+     spec:
+       awsBillingReportDataSource:
+         bucket: gmeteringbucket
+         enabled: true
+         region: ap-southeast-1
```
Comment (timflannagan) [inline modules/metering-troubleshooting.adoc:163]: Should we remove this configuration? Would that be confusing if we explicitly document you need to configure storage, but omit those requisite properties from a MeteringConfig example?

File: modules/metering-troubleshooting.adoc [hunk 5]
```diff
@@ -10,10 +10,173 @@ A common issue with metering is pods failing to start. Pods might fail to start
...
++
+.Example `MeteringConfig` resource
+[source,yaml]
+----
+  kind: MeteringConfig
+  metadata:
+  name: operator-metering
+   namespace: operator-metering
+  spec:
+   openshift-reporting:
```
Comment (timflannagan) [inline modules/metering-troubleshooting.adoc:158]: Hmm, this looks off. If we're configuring resource requests/limits for the reporting operator, then the top-level section would be `spec.reporting-operator` and not `spec.openshift-reporting`.

File: modules/metering-troubleshooting.adoc [hunk 6]
```diff
@@ -10,10 +10,173 @@ A common issue with metering is pods failing to start. Pods might fail to start
...
+       resources: <1>
+         limits:
+           cpu: 1
+           memory: 750Mi
+         requests:
+           cpu: 500m
+           memory: 500Mi
+  ...
+----
+<1> Add or increase `memory` limits within the `resources` field of the `MeteringConfig` resource.
```
Comment (timflannagan) [inline modules/metering-troubleshooting.adoc:173]: Is there a reason we need to code bracket `memory` here?

File: modules/metering-troubleshooting.adoc [hunk 7]
```diff
@@ -10,10 +10,173 @@ A common issue with metering is pods failing to start. Pods might fail to start
...
++
+[NOTE]
+====
+If memory limits are double the recommended defaults and there continue to be numerous OOM killed events, this might indicate that a different issue is causing the reports to be in a pending state.
```
Comment (timflannagan) [inline modules/metering-troubleshooting.adoc:177]: This is also likely fine, but our default recommendations are almost purely speculative and haven't been revisited (or even properly scale tested) in a while to feel confident with documenting this behavior.

File: modules/metering-troubleshooting.adoc [hunk 8]
```diff
@@ -10,10 +10,173 @@ A common issue with metering is pods failing to start. Pods might fail to start
...
+   Environment:
+...
+----
+<1> The current memory limit for the `reporting-operator` pod.
+
+. Edit the `MeteringConfig` resource to update the memory limit:
++
+[source,terminal]
+----
+$ oc edit meteringconfig/operator-metering
```
Comment (timflannagan) [inline modules/metering-troubleshooting.adoc:147]: I think we document a different name than `operator-metering` for the MeteringConfig custom resource. Do we want to also emulate that here, or does that detail of the naming not mean much as long as it's consistent throughout this particular troubleshooting documentation?

File: modules/metering-troubleshooting.adoc (hunk 4 above)
Comment (lbarbeevargas) [reply modules/metering-troubleshooting.adoc:163]: @timflannagan 
Can that section be more generic for the example? 
```
       BillingReportDataSource:
         bucket: meteringbucket
         enabled: true
         region: southeast-1
```

File: modules/metering-troubleshooting.adoc [hunk 9]
```diff
@@ -10,10 +10,173 @@ A common issue with metering is pods failing to start. Pods might fail to start
...
+   Host Ports:     0/TCP, 0/TCP, 0/TCP
+   State:          Running
+     Started:      Thu, 03 Dec 2020 20:59:45 +1000
+   Last State:     Terminated
+     Reason:       OOMKilled <1>
+     Exit Code:    137
+     Started:      Thu, 03 Dec 2020 20:38:05 +1000
+     Finished:     Thu, 03 Dec 2020 20:59:43 +1000
+----
+<1> The `reporting-operator` pod was terminated due to OOM kill.
```
Comment (timflannagan) [inline modules/metering-troubleshooting.adoc:102]: `reporting-operator` -> Reporting Operator?

Or, copy-and-paste the pod from the previous section?

File: modules/metering-troubleshooting.adoc [hunk 10]
```diff
@@ -10,10 +10,173 @@ A common issue with metering is pods failing to start. Pods might fail to start
...
+
+[discrete]
+[id="metering-check-and-increase-memory-limits_{context}"]
+=== Increasing the reporting-operator pod memory limit
+
+If you are experiencing an increase in pod restarts and OOM kill events, you can check the current memory limit set for the `reporting-operator` pod. Increasing the memory limit allows the `reporting-operator` pod to update the report data sources. If necessary, increase the memory limit in your `MeteringConfig` resource by 25-50%.
```
Comment (timflannagan) [inline modules/metering-troubleshooting.adoc:108]: Same feedback here after looking over your PR from this morning. Do we want to substitute `reporting-operator` -> Reporting Operator?

File: modules/metering-troubleshooting.adoc [hunk 11]
```diff
@@ -10,10 +10,173 @@ A common issue with metering is pods failing to start. Pods might fail to start
...
+[source,terminal]
+----
+$ oc describe pod reporting-operator-67d6f57c56-79mrt
+----
++
+.Example output
+[source,terminal]
+----
+Name:         reporting-operator-67d6f57c56-79mrt
+Namespace:    operator-metering
```
Comment (timflannagan) [inline modules/metering-troubleshooting.adoc:123]:
```suggestion
Namespace:    openshift-metering
```

File: modules/metering-troubleshooting.adoc [hunk 12]
```diff
@@ -10,10 +10,173 @@ A common issue with metering is pods failing to start. Pods might fail to start
...
+[source,terminal]
+----
+$ oc describe pod/reporting-operator-84f7c9b7b6-fr697
+----
++
+.Example output
+[source,terminal]
+----
+Name:         reporting-operator-84f7c9b7b6-fr697
+Namespace:    operator-metering
```
Comment (timflannagan) [inline modules/metering-troubleshooting.adoc:88]:
```suggestion
Namespace:    openshift-metering
```

File: modules/metering-troubleshooting.adoc [hunk 13]
```diff
@@ -10,10 +10,173 @@ A common issue with metering is pods failing to start. Pods might fail to start
...
 
+To troubleshoot issues due to a lack of compute resources, check the following within the `operator-metering` namespace.
+
+.Procedure
+
+. Check for metering `Report` resources that fail to complete and show the status of `ReportingPeriodUnmetDependencies`:
++
+[source,terminal]
+----
+$ oc get reports
```
Comment (timflannagan) [inline modules/metering-troubleshooting.adoc:27]: We may need to call out the need that if you're running commands like `oc get reports` and not supplying the namespace (`-n/--namespace`) parameter, we'll be querying for these resources in whatever the current kube config context is, and not the openshift-metering namespace.

I don't think this is a massive detail, but a quick prerequisite to this procedural section would just be adding the following command:

```oc project openshift-metering```

So you're in the current context when querying for metering-related resources.

File: modules/metering-troubleshooting.adoc (hunk 6 above)
Comment (lbarbeevargas) [reply modules/metering-troubleshooting.adoc:173]: @timflannagan No specific need. I had code bracketed `memory` to refer to the two `memory` fields in the `resources` section. But if that's confusing, I can removed the backticks.

File: modules/metering-troubleshooting.adoc (hunk 4 above)
Comment (timflannagan) [reply modules/metering-troubleshooting.adoc:163]: I realize we're configuring AWS billing here, and not the storage configuration, so we can omit this section entirely.

File: modules/metering-troubleshooting.adoc (hunk 5 above)
Comment (timflannagan) [reply modules/metering-troubleshooting.adoc:158]: This would be an example of configuring requests/limits for the reporting operator: https://github.com/kube-reporting/metering-operator/blob/master/manifests/metering-config/default-resource-limits.yaml#L5-L14.

File: modules/metering-troubleshooting.adoc (hunk 11 above)
Comment (lbarbeevargas) [reply modules/metering-troubleshooting.adoc:123]: @timflannagan Was going to ask if I should go through the rest of this updated section and s/operator-metering/openshift-metering/ 👍

File: modules/metering-troubleshooting.adoc (hunk 6 above)
Comment (timflannagan) [reply modules/metering-troubleshooting.adoc:173]: I think I'd prefer if we removed/reworded:

"Add or increase the memory resource limits within the `resources` field of the `MeteringConfig` resource."

^ That sounds a bit wordy though, so maybe we can just leave it alone for now.

File: modules/metering-troubleshooting.adoc [hunk 14]
```diff
@@ -10,10 +10,173 @@ A common issue with metering is pods failing to start. Pods might fail to start
 [id="metering-not-enough-compute-resources_{context}"]
 == Not enough compute resources
 
-A common issue when installing or running metering is lack of compute resources. Ensure that metering is allocated the minimum resource requirements described in the installation prerequisites.
+A common issue when installing or running metering is a lack of compute resources. As the cluster grows and more reports are created, the `reporting-operator` pod requires more memory. If memory usage reaches the pod limit, the cluster terminates the pod with an `OOMKilled` status. Ensure that metering is allocated the minimum resource requirements described in the installation prerequisites.
```
Comment (adellape) [inline modules/metering-troubleshooting.adoc:13]: Dunno if worth it, but wondering if there's a good way to sneak in the definition of "OOM" here in this first related instance (didn't see it mentioned explicitly in IBM SG or RH supplemental guide). Maybe something like:

> If memory usage reaches the pod limit, the cluster terminates the pod with an `OOMKilled` status due to an out of memory (OOM) event.

or

> If memory usage reaches the pod limit, the cluster considers the pod out of memory (OOM) and terminates it with an `OOMKilled` status.

Then the later, non-marked-up instances of the "OOM" acronym will be covered.

Comment (lbarbeevargas) [reply modules/metering-troubleshooting.adoc:13]: Great suggestion, thank you.

File: modules/metering-troubleshooting.adoc (hunk 10 above)
Comment (lbarbeevargas) [reply modules/metering-troubleshooting.adoc:108]: @timflannagan @adellape I think for general usage/mention, "Reporting Operator pod" works. 👍  

When saying to check specific information regarding the pod, WDYT of the following example:

> Check the current memory limits of the `reporting-operator` `Pod` resource:

File: modules/metering-troubleshooting.adoc (hunk 8 above)
Comment (lbarbeevargas) [reply modules/metering-troubleshooting.adoc:147]: Looking at the [Common configuration options](https://docs.openshift.com/container-platform/4.7/metering/configuring_metering/metering-common-config-options.html) section, we use `operator-metering`.

File: modules/metering-troubleshooting.adoc (hunk 7 above)
Comment (lbarbeevargas) [reply modules/metering-troubleshooting.adoc:177]: @timflannagan Attempt to make this more generic than "double the recommended defaults":

> If there continue to be numerous OOM killed events after memory limits are increased, this might indicate that a different issue is causing the reports to be in a pending state.

**Comment by pruan-rht** [comment]
LGTM

File: modules/metering-troubleshooting.adoc (hunk 10 above)
Comment (timflannagan) [reply modules/metering-troubleshooting.adoc]: That sounds fine to me, but I thought we weren't capitalizing Pod in any circumstances anymore?

**Review by timflannagan — APPROVED** [review]

File: modules/metering-troubleshooting.adoc (hunk 10 above)
Comment (lbarbeevargas) [reply modules/metering-troubleshooting.adoc]: Pod can be capitalized if we're referring to a specific API object. So it's in backticks followed by the type of object. In this case, "`Pod` resource".
